_posts/2018-04-15-customreaders.md

```diff
@@ -0,0 +1,223 @@
+    
+
+    /home/richard/miniconda2/envs/mydev/lib/python2.7/site-packages/MDAnalysis/core/universe.pyc in __init__(self, *args, **kwargs)
+        268             else:
+        269                 coordinatefile = args[1:]
+    --> 270             self.load_new(coordinatefile, **kwargs)
+        271 
+        272         # Check for guess_bonds
+
+
+    /home/richard/miniconda2/envs/mydev/lib/python2.7/site-packages/MDAnalysis/core/universe.pyc in load_new(self, filename, format, in_memory, **kwargs)
+        416         kwargs['n_atoms'] = self.atoms.n_atoms
+        417 
+    --> 418         self.trajectory = reader(filename, **kwargs)
+        419         if self.trajectory.n_atoms != len(self.atoms):
+        420             raise ValueError("The topology and {form} trajectory files don't"
+
+
+    /home/richard/miniconda2/envs/mydev/lib/python2.7/site-packages/MDAnalysis/coordinates/XYZ.pyc in __init__(self, filename, **kwargs)
+        309         # (Also cannot just use seek() or reset() because that would break
+        310         # with urllib2.urlopen() streams)
+    --> 311         self._read_next_timestep()
+        312 
+        313     @property
+
+
+    /home/richard/miniconda2/envs/mydev/lib/python2.7/site-packages/MDAnalysis/coordinates/XYZ.pyc in _read_next_timestep(self, ts)
+        369             return ts
+        370         except (ValueError, IndexError) as err:
+    --> 371             raise EOFError(err)
+        372 
+        373     def _reopen(self):
+
+
+    EOFError: could not convert string to float: red
+
+
+Here, as expected, the XYZReader chokes on the word 'red' where it expects a floating point number.
+
+We can rewrite the XYZReader to anticipate the extra column:
+
+
+```python
+import numpy as np
+from MDAnalysis.coordinates.XYZ import XYZReader
+from MDAnalysis.topology.XYZParser import XYZParser
+
+class XYZCReader(XYZReader):
+    format = 'XYZC'
+    
+    def _read_next_timestep(self, ts=None):
+        # check that the timestep object exists
+        if ts is None:
+            ts = self.ts
+
+        f = self.xyzfile
+
+        try:
+            # we assume that there are only two header lines per frame
+            f.readline()
+            f.readline()
+            for i in range(self.n_atoms):
+                self.ts._pos[i] = np.float32(f.readline().split()[2:5])
+            ts.frame += 1
+            return ts
+        except (ValueError, IndexError) as err:
+            raise EOFError(err)
+            
+class XYZCParser(XYZParser):
+    format = 'XYZC'
+```
+
+
+```python
+u = mda.Universe('myfile.xyzc')
+```
+
+
+```python
+print u.atoms.names
+print u.atoms.positions
+```
+
+    ['C' 'H' 'H' 'C' 'H' 'H' 'O']
+    [[ 1.   1.   1. ]
+     [ 1.   1.5  1.5]
+     [ 1.   0.5  0.5]
+     [ 2.   2.   2. ]
+     [ 2.   2.5  2.5]
+     [ 2.   1.5  1.5]
+     [ 3.   3.   3. ]]
+
+
+Now it seems to work.
+
+But what about our colour field?
```

Comment:
Show a example that the attribute doesn't exist.  Then

`To add the color field to our topology we need to teach MDAnalysis about the color attribute in general and our new XYZC parser to read it.`